gdal/frmts/fits/fitsdataset.cpp

```diff
@@ -537,35 +657,50 @@ GDALDataset *FITSDataset::Create( const char* pszFilename,
   // removed on 2002-07-02 by Simon Perkins because they introduced
   // excessive complications and didn't really fit into the GDAL
   // paradigm.
+  // 2018 - BZERO BSCALE keywords are now set using SetScale() and
+  // SetOffset() functions 
 
-  // Create the file - to force creation, we prepend the name with '!'
-  char* extFilename = new char[strlen(pszFilename) + 10];  // 10 for margin!
-  snprintf(extFilename, strlen(pszFilename) + 10, "!%s", pszFilename);
-  fitsfile* hFITS = nullptr;
-  fits_create_file(&hFITS, extFilename, &status);
-  delete[] extFilename;
-  if (status) {
-    CPLError(CE_Failure, CPLE_AppDefined,
-             "Couldn't create FITS file %s (%d).\n", pszFilename, status);
-    return nullptr;
+  if( nXSize < 1 || nYSize < 1 || nBands < 1 )  {
+        CPLError(
+            CE_Failure, CPLE_AppDefined,
+            "Attempt to create %dx%dx%d raster FITS file, but width, height and bands"
+            " must be positive.",
+            nXSize, nYSize, nBands );
+
+        return nullptr;
   }
 
   // Determine FITS type of image
   int bitpix;
-  if (eType == GDT_Byte)
+  if (eType == GDT_Byte) {
     bitpix = BYTE_IMG;
-  else if (eType == GDT_Int16)
+  } else if (eType == GDT_UInt16) {
+    bitpix = USHORT_IMG;
+  } else if (eType == GDT_Int16) {
     bitpix = SHORT_IMG;
-  else if (eType == GDT_Int32)
+  } else if (eType == GDT_UInt32) {
+    bitpix = ULONG_IMG;
+  } else if (eType == GDT_Int32) {
     bitpix = LONG_IMG;
-  else if (eType == GDT_Float32)
+  } else if (eType == GDT_Float32)
     bitpix = FLOAT_IMG;
   else if (eType == GDT_Float64)
     bitpix = DOUBLE_IMG;
   else {
     CPLError(CE_Failure, CPLE_AppDefined,
              "GDALDataType (%d) unsupported for FITS", eType);
-    fits_close_file(hFITS, &status);
+    return nullptr;
+  }
+
+  // Create the file - to force creation, we prepend the name with '!'
+  char* extFilename = new char[strlen(pszFilename) + 10];  // 10 for margin!
```

Comment:
Could be simplified by doing
```
CPLString extFilename("!");
extFilename += pszFilename;
```

Reply:
Fixed in https://github.com/OSGeo/gdal/pull/1298/commits/202323471002d33a3acfcc0549fe8312d9e2d680